### PR TITLE
fix(platform): remove hardcoded content densities

### LINKED ISSUE #7573: bug(core|platform): hardcoded content densities
#### Is this a bug, enhancement, or feature request?
bug
#### Briefly describe your proposal.
Some components have hardcoded content densities. 

Instead the following behavior should be used:
- components should attempt to pull content density from the application level. If it's not possible, fallback to default value.
- should be able to override content density for component

### PR DESCRIPTION
**Pull request by dmitry-stepanenko**
## Related Issue(s)

closes #7573

## Description
Some components have hardcoded content densities.

Instead the following behavior should be used:

components should attempt to pull content density from the application level. If it's not possible, fallback to default value.
should be able to override content density for component

In this PR fixed content densities for:
- platform-table
- platform-upload-collection
- platform-value-help-dialog

Additionally fixed an issue with `condensed` content density: most of the platform components were ignoring it and displaying themselves as `cozy`. Now `compact` will be used

##### During Implementation

1. Visual Testing:

-   [x] visual misalignments/updates
-   [x] check Light/Dark/HCB/HCW themes
-   [x] RTL/LTR - proper rendering and labeling
-   [x] responsiveness(resize)
-   [x] Content Density (Cozy/Compact/(Condensed))
-   [x] States - hover/disabled/focused/active/on click/selected/selected hover/press state
-   [x] Interaction/Animation - open/close, expand/collapse, add/remove, check/uncheck
-   [x] Mouse vs. Keyboard support
-   [x] Text Truncation

2. API and functional correctness

-   [x] check for console logs (warnings, errors)
-   [x] API boundary values
-   [n/a] different combinations of components - free style
-   [x] change the API values during testing

3. Documentation and Example validations

-   [x] missing API documentation or it is not understandable
-   [n/a] poor examples
-   [n/a] Stackblitz works for all examples

4. Accessibility testing
5. Browser Testing - Edge, Safari, Chrome, Firefox

##### PR Quality

-   [x] the commit message(s) follows the guideline:
        https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
-   [n/a] tests for the changes that have been done
-   [x] all items on the PR Review Checklist are addressed :
        https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
-   [n/a] Run npm run build-pack-library and test in external application
-   [n/a] update `README.md`
-   [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

### CODE REVIEW

File: libs/platform/src/lib/table/components/reset-button/reset-button.component.ts [hunk 1]
```diff
@@ -29,5 +29,7 @@ export const RESETTABLE_TOKEN = new InjectionToken<Resettable>('Resettable');
     changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class ResetButtonComponent {
+    @Input() compact: boolean;
```
Comment (InnaAtanasova) [inline libs/platform/src/lib/table/components/reset-button/reset-button.component.ts:32]: missing typedocs

File: libs/platform/src/lib/table/components/table-p13-dialog/filtering/filter-rule.component.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ import { FilterRule } from './filtering.model';
 })
 export class FilterRuleComponent implements OnDestroy {
     @Input() rule: FilterRule;
+    @Input() compact: boolean;
```
Comment (InnaAtanasova) [inline libs/platform/src/lib/table/components/table-p13-dialog/filtering/filter-rule.component.ts:18]: missing typedocs

File: libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts [hunk 3]
```diff
@@ -286,10 +306,30 @@ export class UploadCollectionComponent {
...
-        private readonly _filesValidatorService: FilesValidatorService
-    ) {}
+        private readonly _filesValidatorService: FilesValidatorService,
+        private readonly _cdr: ChangeDetectorRef,
+        @Optional() private readonly _contentDensityService: ContentDensityService
+    ) {
+        this._trackContentDensityChanges();
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
```
Comment (InnaAtanasova) [inline libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts:321]: should be hidden

File: libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts [hunk 4]
```diff
@@ -286,10 +306,30 @@ export class UploadCollectionComponent {
...
+        if (changes.contentDensity) {
+            this._contentDensityManuallySet = true;
+        }
+    }
+
+    ngOnDestroy(): void {
```
Comment (InnaAtanasova) [inline libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts:327]: should be hidden

File: libs/platform/src/lib/table/components/reset-button/reset-button.component.ts (hunk 1 above)
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/table/components/reset-button/reset-button.component.ts:32]: done

File: libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts (hunk 3 above)
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts:322]: done

File: libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts (hunk 4 above)
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/upload-collection/upload-collection/upload-collection.component.ts:329]: done

File: libs/platform/src/lib/value-help-dialog/components/define-tab/define-tab.component.html [hunk 5]
```diff
@@ -23,7 +23,7 @@
                 colGrow
                 [class.fd-select-custom-class]="false"
                 #sttr
-                [compact]="true"
+                [compact]="isCompact"
```
Comment (platon-rov) [inline libs/platform/src/lib/value-help-dialog/components/define-tab/define-tab.component.html:26]: In all other components `contentDensity !== 'cozy'` used in template, and here getter is provided, let's unify the approach :)

File: libs/platform/src/lib/value-help-dialog/value-help-dialog/value-help-dialog.component.html [hunk 6]
```diff
@@ -43,12 +43,12 @@ <h5 fd-title class="selected-and-condition-title" [id]="id + '-selected-item'">
                     <div class="fdp-value-help-dialog__tokens-wrapper">
                         <div class="fdp-value-help-dialog__tokens">
                             <!-- Selected items -->
-                            <fd-tokenizer [tokenizerFocusable]="false" [compact]="true">
+                            <fd-tokenizer [tokenizerFocusable]="false" [compact]="isCompact">
```
Comment (platon-rov) [inline libs/platform/src/lib/value-help-dialog/value-help-dialog/value-help-dialog.component.html:46]: Oh, also getter is used.

File: libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html [hunk 7]
```diff
@@ -37,7 +37,7 @@ <h5 fd-title role="heading" aria-level="5">
                     <fd-checkbox
                         *ngIf="isMultiSelection"
                         [name]="uid + '-row-all'"
-                        [compact]="true"
+                        [compact]="contentTableDensity === _contentDensityOptions.COMPACT"
```
Comment (platon-rov) [inline libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html:40]: `!== 'cozy'` should work?

**Review by platon-rov — COMMENTED** [review]
Great work, nice to see that such library-wide improvements :)

But i noticed that several approaches are used

* `contentDensity !== 'cozy'`
* getter
* pipes (in table & subcomponents)

Maybe we should unify the approach thru all the components? Since putting getters for every component doesn't make sense and having hardcoded checks `!== 'cozy'` also may be not the best solution i think pipes is pretty good approach here, i would like to vote to make it shared & use everywhere.

**Comment by dmitry-stepanenko** [comment]
@platon-rov sounds reasonable, switched to using pipe everywhere

File: libs/core/src/lib/utils/services/content-density.service.ts [hunk 8]
```diff
@@ -15,15 +16,16 @@ export const DEFAULT_CONTENT_DENSITY: ContentDensity = 'cozy';
  */
 @Injectable()
 export class ContentDensityService {
-    private _contentDensity: BehaviorSubject<ContentDensity> = new BehaviorSubject(DEFAULT_CONTENT_DENSITY);
+    /** Content Density BehaviourSubject */
+    readonly contentDensity = new BehaviorSubject(DEFAULT_CONTENT_DENSITY);
```
Comment (platon-rov) [inline libs/core/src/lib/utils/services/content-density.service.ts:20]: Previously it was private and only getter was revealed, now both of them are revealed, is there a reason for that?

File: libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html (hunk 7 above)
Comment (platon-rov) [reply libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html:40]: Seems like it's still actual. Pipe should be used.

**Review by platon-rov — APPROVED** [review]
Good!

File: libs/core/src/lib/utils/services/content-density.service.ts (hunk 8 above)
Comment (dmitry-stepanenko) [reply libs/core/src/lib/utils/services/content-density.service.ts:20]: If you look closer, it's exactly the same: previously it was a getter for behavior subject, which served a purpose to make it read-only. This small refactoring just makes the syntax lighter, nothing is added, nothing is lost

File: libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html (hunk 7 above)
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/value-help-dialog/components/select-tab/select-tab.component.html]: yes, thanks. fixed

File: libs/core/src/lib/utils/services/content-density.service.ts (hunk 8 above)
Comment (platon-rov) [reply libs/core/src/lib/utils/services/content-density.service.ts:20]: Yea, i see, thought that previously getter was returning an `Observable`.

**Pull request merged**